rfd/0165-ssh-connection-handover.md

```diff
@@ -0,0 +1,53 @@
+
+The server-side of the hand-over connection will first read the remote IP address, then proceed operating with the connection as if a client with the received remote IP address had just attempted to resume a connection with the token associated to the hand-over connection, by replying with a "success" tag if the remote address matches the one of the resumable connection, a "bad address" tag otherwise, or a "not found" tag if the connection had been closed in the meantime.
+
+### Graceful restart by default
+
+With this change we're able to treat concurrent instances of Teleport running on the same data directory (and thus "hand-over registry") almost interchangeably, with regards to SSH connections; that's effectively what happens during a graceful restart initiated by SIGHUP: Teleport executes a new Teleport process launched from disk, with the same configuration, passing it a copy of all the listeners (if any) used by the configured services, then waits for it to signal that Teleport has started successfully, then waits for existing connections to complete just like during a graceful shutdown. Similarly, SIGUSR2 process forking (which shouldn't be used) and internal process restart due to host CA rotation (which we should get rid of) also end up running multiple copies of Teleport with the same configuration and credentials. SSH connection hand-over allows connection resumption to work in all such cases.
+
+The current behavior of the `systemd-unit-upgrader` script is to initiate a systemd service restart after upgrading the Teleport package on disk. A systemd service restart (by default) sends a termination signal to the "main process" of the unit, then after it exits (or after a timeout) the whole cgroup of the service is killed, and a new copy of the service is started. We're going to change the unit upgrader to only restart in the presence of upgrades marked "critical" (as those generally indicate security-sensitive upgrades, and we should not risk running vulnerable binaries for the sake of not having to reopen connections), and to reload the service (sending Teleport a SIGHUP) in other cases. With this change, a (non-critical) Teleport upgrade should result in completely unaffected SSH sessions, even long lived ones, as well as allowing sessions for other protocols to run until completion. An optional flag can be set in the configuration for `systemd-unit-upgrader` to disable reloading and always do a clean restart.
```

Comment:
> - adding a deadline so a replaced process quits at some point  

In case we decide to implement this (now or in the future), a draft PR to add a deadline to graceful shutdown and graceful reloading is up at #38544 - a very small change.

If a server is used often enough for "upgrade pileup" to become an issue, I suspect that it's also used by users who will be livid about the arbitrary termination of their SSH session.

The `disconnect_expired_cert` role option and cluster auth preference option will still apply to connections during reload, so we could leave the choice to the cluster admins rather than enforce a limit in the shutdown handling.

> - limiting the process concurrency to 2 or 3

I don't even know where to begin to implement this in a way that makes sense, since the process that should exit (the oldest one) will be blissfully unaware of its disowned grandchildren 4 processes down the chain - especially if some Teleport in the middle has already exited and its children were reparented to init 😬

Reply:
@russjones Can we revisit the lack of timeout on reload/shutdown at a later time if it turns out to be an actual problem, or should we address it before we start automatically reloading? What's your take on this?

Reply:
Something that maybe wasn't clarified yet in this thread is that Teleport agent processes will only accumulate if there's at least one connection open in each of them; processes that are not actually driving a SSH connection themselves will exit, even if a parent process is still running.

Reply:
Would implementation of the https://github.com/gravitational/cloud/pull/7716 solve the problem of timeout? That seems like a reasonable fix to me.

Reply:
Yes, that's why I prepared it, but it's a change in behavior that will result in connections being summarily cut after the reload - after 30 hours (the maximum login credential TTL), in that implementation.

If that's fine I'll update the RFD to mention it and I'll go ahead and get that reviewed and merged.

Reply:
Seems reasonable since a session can't last longer than the max TTL on the certificate anyway right?

Reply:
It's also possible to generate overlong user credentials with `tctl auth sign` or similar.

30 hours is still probably beyond most reasonable expectations of maintaining individual sessions open tho.

Reply:
It's not perfect, but I think it's reasonable.

Solves the problem of runaway process and we should guide people not to use `tctl auth sign` and use Machine ID instead.

I am fine addressing the case @rosstimothy mentioned (creating within that 30 hour window, then keeping session around longer) if it becomes an issue.

We should document this behavior and make it clear to administrators.

What do you guys think?

Reply:
Well, I've merged the PR in master already. 🤭

I think that a 30 hour timeout is generous enough to not really affect anything, and if some user really needs the timeout to be configurable we'll make it configurable.